lib/timers.js

```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
   var diff, first, threw;
   while (first = L.peek(list)) {
     diff = now - first._idleStart;
+    // This handles the case of a timer that was created within a timers
+    // callback with the same timeout value. For instance, when processing the
+    // timer that would call `bar` in such code:
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 0) }, 0);
+    //
+    // or
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);
+    //
+    // We want to make sure that newly added timer fires in the next turn of the
+    // event loop at the earliest. So even if it's already expired now,
+    // reschedule it to fire later.
```

Comment:
I'm not sure this is actually a bug. Shouldn't the timer be scheduled in `setImmediate()` if you want to guarantee this?

Also, how does this not only apply to 0ms timeout timers?

> `setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);`

In this example, the 2nd timeout should be caught by https://github.com/nodejs/node/blob/master/lib/timers.js#L68